File: components/shell.py

from components.subshell import subshell

class shell:
    def __init__(self, energy_level):
        self.subshells = [subshell(-1, -1)]
        self.energy_level = energy_level
        for i in range(1, min(energy_level, 4)+1):
            self.subshells.append(subshell(i, i))
    
    def occupied(self, which_level):
        return self.subshells[which_level].occupied

    def howMany(self):
        sum = 0
        for current_subshell in range(1, len(self.subshells)):
            sum += self.occupied(current_subshell)
        return sum
    
    def valence(self):
        return self.occupied(self.energy_level)
        

File: shell/shell.cpp
#include "shell.h"
shell::shell(int k) {
  energyLevel = k;
  // there are only 4 maxmimum subshells possible, and so the min() function takes care of this for smaller and larger i
  for (int i = 1; i <= min(k, 4); i++) 
    subshells.push_back(subshell(i, i));
}
// get sum for each subshell
int shell::occupied(int whichLevel) { 
  return subshells[whichLevel].occupied;
}
// sum of all the total electrons in all subshells of a particular shell 
int shell::howMany() {  
  int sum = 0;
  for (int i = 1; i < subshells.size(); i++)
    sum += occupied(i);
  return sum;
}
// returns number of electrons in the valence shell 
int shell::valence() { 
  return occupied(energyLevel);
}